src/pages/Profile.js

```diff
@@ -154,6 +155,114 @@ const NihLink = ({ nihToken }) => {
+  const decodeProvider = state => state ? JSON.parse(atob(state)).provider : ''
+
+  const extractToken = (provider, { state, code }) => {
+    const extractedProvider = decodeProvider(state)
+    return extractedProvider && provider === extractedProvider ? code : undefined
+  }
+
+  const queryParams = qs.parse(window.location.search, { ignoreQueryPrefix: true })
+  const token = extractToken(provider, queryParams)
+  const redirectUrl = `${window.location.origin}/${Nav.getLink('fence-callback')}`
+
+  /*
+   * Hooks
+   */
+  const [{ username, issued_at: issuedAt }, setStatus] = useState({})
+  const [href, setHref] = useState(undefined)
+  const [isLoadingStatus, setIsLoadingStatus] = useState(false)
+  const [isLoadingAuthUrl, setIsLoadingAuthUrl] = useState(false)
+  const [isLinking, setIsLinking] = useState(false)
+  const signal = useCancellation()
+
+  const { User } = Ajax(signal)
+
+  Utils.useOnMount(() => {
+    const linkFenceAccount = _.flow(
+      withErrorReporting('Error linking NIH account'),
+      Utils.withBusyState(setIsLinking)
+    )(async () => {
+      setStatus(await User.linkFenceAccount(provider, token, redirectUrl))
+    })
+
+    if (token) {
+      const profileLink = `/${Nav.getLink('profile')}`
+      window.history.replaceState({}, '', profileLink)
+      linkFenceAccount()
+    }
+  })
+
+  Utils.useOnMount(() => {
+    const loadFenceStatus = _.flow(
+      withErrorReporting(`Error loading status for ${displayName}`),
+      Utils.withBusyState(setIsLoadingStatus)
+    )(async () => {
+      try {
+        setStatus(await User.getFenceStatus(provider))
+      } catch (error) {
+        if (error.status === 404) {
+          setStatus({})
+        } else {
+          throw error
+        }
+      }
+    })
+
+    if (!token) {
+      loadFenceStatus()
+    }
+  })
+
+  Utils.useOnMount(() => {
+    const loadAuthUrl = _.flow(
+      withErrorReporting(`Error loading Fence link`),
+      Utils.withBusyState(setIsLoadingAuthUrl)
+    )(async () => {
+      const result = await User.getFenceAuthUrl(provider, redirectUrl)
+      setHref(result.url)
+    })
+
+    loadAuthUrl()
+  })
+
+  /*
+   * Render helpers
+   */
+  const makeAccountLinkLink = () => {
+    return href && link({ href, style: { display: 'flex', alignItems: 'center' } }, [
+      'Log-In to Framework Services to link your account',
+      icon('pop-out', { size: 12, style: { marginLeft: '0.5rem' } })
+    ])
+  }
+
+  /*
+   * Render
+   */
+  const isBusy = () => isLoadingStatus || isLoadingAuthUrl || isLinking
+  const expireTime = new Date(issuedAt).getTime() + (1000 * 60 * 60 * 24 * 30)
+
+  return div({ style: { marginBottom: '1rem' } }, [
+    div({ style: styles.form.title }, [displayName]),
+    isBusy() && div([spinner(), 'Loading account status...']),
+    !isBusy() && h(Fragment, [
+      !username && makeAccountLinkLink(),
+      !!username && div({ style: { display: 'flex', flexDirection: 'column', width: '33rem' } }, [
```

Comment:
nit: for these mutually exclusive items, a ternary might be clearer.

Reply:
I am sympathetic to the repetition of the condition as this is currently written. However, I've found ternary to be less readable when the possible outcomes span multiple lines; it's difficult to visually spot the `:`. I've also seen a ternary be confusing when the condition includes `&&` because, at first glance, it looks like the if-condition-then-render-a-thing pattern instead of a ternary.

One option is to extract the `div(...)` into a render helper so a ternary would fit on one line. Another would be `Utils.cond`:
```
Utils.cond(
  [!username, () => renderFenceLink()]
  () => div({...}, [
    ...
  ])
)
```
`Utils.cond` feels like a heavy hammer for this, but the single condition with a default value expresses the if/else nature.

Or just tell me to get used to reading ternary here and I'll do that.

Reply:
Chatted in person. Decided this is fine... at least until we come to consensus on a preferred style for these cases.